File: src/main/scala/committee/nova/skillful/network/handler/NetworkHandler.scala
package committee.nova.skillful.network.handler

import committee.nova.skillful.Constants
import committee.nova.skillful.network.message.{InfoClearMessage, SkillsSyncMessage}
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.network.simple.SimpleChannel
import net.minecraftforge.fml.network.{NetworkEvent, NetworkRegistry}

import java.util.function.Supplier

object NetworkHandler {
  final val version = "1.0"
  var instance: SimpleChannel = _
  private var id: Int = 0

  private def nextId: Int = {
    id += 1
    id
  }

  def registerMessage(): Unit = {
    instance = NetworkRegistry.newSimpleChannel(
      new ResourceLocation(Constants.MODID, Constants.MODID),
      () => version,
      v => v.equals(version),
      v => v.equals(version)
    )
    instance.messageBuilder(classOf[InfoClearMessage], nextId)
      .encoder((_, _) => {})
      .decoder(_ => new InfoClearMessage)
      .consumer((m: InfoClearMessage, ctx: Supplier[NetworkEvent.Context]) => m.handler(ctx))
      .add()
    instance.messageBuilder(classOf[SkillsSyncMessage], nextId)
      .encoder((m, p) => m.toBytes(p))
      .decoder(p => new SkillsSyncMessage(p))
      .consumer((m: SkillsSyncMessage, ctx: Supplier[NetworkEvent.Context]) => m.handler(ctx))
      .add()
  }
}

File: src/main/java/committee/nova/skillful/common/network/handler/NetworkHandler.java
package committee.nova.skillful.common.network.handler;

import committee.nova.skillful.Skillful;
import committee.nova.skillful.common.network.msg.SyncSingleSkillMsg;
import committee.nova.skillful.common.network.msg.SyncSkillsMsg;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.network.NetworkRegistry;
import net.minecraftforge.network.simple.SimpleChannel;

public class NetworkHandler {
    private static SimpleChannel instance;
    public static final String VERSION = "1.0";
    private static int id = 0;

    public static void registerMsg() {
        instance = NetworkRegistry.newSimpleChannel(
                new ResourceLocation(Skillful.MODID, "msg"),
                () -> VERSION,
                VERSION::equals,
                VERSION::equals
        );
        instance.messageBuilder(SyncSingleSkillMsg.class, nextId())
                .encoder(SyncSingleSkillMsg::toBytes)
                .decoder(SyncSingleSkillMsg::new)
                .consumer(SyncSingleSkillMsg::handler)
                .add();
        instance.messageBuilder(SyncSkillsMsg.class, nextId())
                .encoder(SyncSkillsMsg::toBytes)
                .decoder(SyncSkillsMsg::new)
                .consumer(SyncSkillsMsg::handler)
                .add();
    }

    public static int nextId() {
        return id++;
    }

    public static SimpleChannel getInstance() {
        return instance;
    }
}
